File: go/internal/tools/linode_instance_backups.go
package tools

import (
	"context"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/chadit/LinodeMCP/go/internal/config"
	"github.com/chadit/LinodeMCP/go/internal/linode"
	"github.com/chadit/LinodeMCP/go/internal/profiles"
	"github.com/chadit/LinodeMCP/go/internal/toolschemas"
	"github.com/chadit/LinodeMCP/go/internal/twostage"
)

// NewLinodeInstanceBackupListTool creates a tool for listing all backups for a Linode instance.
func NewLinodeInstanceBackupListTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newToolWithHandler(
		cfg,
		"linode_instance_backup_list",
		"Lists all backups for a Linode instance, including automatic backups and manual snapshots.",
		[]mcp.ToolOption{
			mcp.WithNumber("linode_id", mcp.Required(),
				mcp.Description("The ID of the Linode instance")),
		},
		handleInstanceBackupsListRequest,
	)

	return tool, profiles.CapRead, handler
}

func handleInstanceBackupsListRequest(ctx context.Context, request *mcp.CallToolRequest, cfg *config.Config) (*mcp.CallToolResult, error) {
	linodeID := request.GetInt("linode_id", 0)
	if linodeID == 0 {
		return mcp.NewToolResultError(ErrLinodeIDRequired.Error()), nil
	}

	client, err := prepareClient(request, cfg)
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	backups, err := client.ListInstanceBackups(ctx, linodeID)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("Failed to list backups for instance %d: %v", linodeID, err)), nil
	}

	return MarshalToolResponse(backups)
}

// NewLinodeInstanceBackupGetTool creates a tool for retrieving a specific backup for a Linode instance.
func NewLinodeInstanceBackupGetTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool := mcp.NewToolWithRawSchema(
		"linode_instance_backup_get",
		"Retrieves details of a specific backup for a Linode instance.",
		toolschemas.Schema("linode.mcp.v1.InstanceBackupGetInput"),
	)

	handler := func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		return handleInstanceBackupGetRequest(ctx, &request, cfg)
	}

	return tool, profiles.CapRead, handler
}

func handleInstanceBackupGetRequest(ctx context.Context, request *mcp.CallToolRequest, cfg *config.Config) (*mcp.CallToolResult, error) {
	linodeID := request.GetInt("linode_id", 0)
	if linodeID == 0 {
		return mcp.NewToolResultError(ErrLinodeIDRequired.Error()), nil
	}

	backupID := request.GetInt("backup_id", 0)
	if backupID == 0 {
		return mcp.NewToolResultError(ErrBackupIDRequired.Error()), nil
	}

	client, err := prepareClient(request, cfg)
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	backup, err := client.GetInstanceBackupProto(ctx, linodeID, backupID)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("Failed to retrieve backup %d for instance %d: %v", backupID, linodeID, err)), nil
	}

	return MarshalProtoToolResponse(backup)
}

// NewLinodeInstanceBackupCreateTool creates a tool for taking a manual snapshot of a Linode instance.
func NewLinodeInstanceBackupCreateTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newToolWithHandler(
		cfg,
		"linode_instance_backup_create",
		"Creates a manual snapshot of a Linode instance. "+
			"WARNING: This overwrites any existing manual snapshot for the instance.",
		[]mcp.ToolOption{
			mcp.WithNumber("linode_id", mcp.Required(),
				mcp.Description("The ID of the Linode instance to snapshot")),
			mcp.WithString("label",
				mcp.Description("Label for the manual snapshot (optional)")),
			mcp.WithBoolean(paramConfirm, mcp.Required(),
				mcp.Description("Must be true to confirm snapshot creation. This overwrites any existing manual snapshot. Ignored when dry_run=true.")),
			mcp.WithBoolean(paramDryRun, mcp.Description(paramDryRunDesc)),
		},
		handleInstanceBackupCreateRequest,
	)

	return tool, profiles.CapWrite, handler
}

func handleInstanceBackupCreateRequest(ctx context.Context, request *mcp.CallToolRequest, cfg *config.Config) (*mcp.CallToolResult, error) {
	linodeID := request.GetInt("linode_id", 0)

	if IsDryRun(request) {
		if linodeID == 0 {
			return mcp.NewToolResultError(ErrLinodeIDRequired.Error()), nil
		}

		return RunDryRunPreview(ctx, request, cfg, "linode_instance_backup_create", httpMethodPost,
			fmt.Sprintf("/linode/instances/%d/backups", linodeID),
			func(ctx context.Context, c *linode.Client) (any, error) { return c.GetInstance(ctx, linodeID) })
	}

	if result := RequireConfirm(request, "This creates a manual snapshot and overwrites any existing one. Set confirm=true to proceed."); result != nil {
		return result, nil
	}

	if linodeID == 0 {
		return mcp.NewToolResultError(ErrLinodeIDRequired.Error()), nil
	}

	client, err := prepareClient(request, cfg)
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	backup, err := client.CreateInstanceBackup(ctx, linodeID, request.GetString("label", ""))
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("Failed to create snapshot for instance %d: %v", linodeID, err)), nil
	}

	response := struct {
		Message string                 `json:"message"`
		Backup  *linode.InstanceBackup `json:"backup"`
	}{
		Message: fmt.Sprintf("Snapshot created for instance %d (backup ID: %d)", linodeID, backup.ID),
		Backup:  backup,
	}

	return MarshalToolResponse(response)
}

// NewLinodeInstanceBackupRestoreTool creates a tool for restoring a backup to a Linode instance.
func NewLinodeInstanceBackupRestoreTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newToolWithHandler(
		cfg,
		"linode_instance_backup_restore",
		"Restores a backup to a Linode instance. "+
			"WARNING: When overwrite is true, this destroys all disks and configs on the target instance.",
		[]mcp.ToolOption{
			mcp.WithNumber("linode_id", mcp.Required(),
				mcp.Description("The ID of the Linode instance that owns the backup")),
			mcp.WithNumber("backup_id", mcp.Required(),
				mcp.Description("The ID of the backup to restore")),
			mcp.WithNumber("target_linode_id", mcp.Required(),
				mcp.Description("The ID of the Linode instance to restore the backup to")),
			mcp.WithBoolean("overwrite",
				mcp.Description("If true, deletes all disks and configs on the target before restoring. Defaults to false.")),
			mcp.WithBoolean(paramConfirm, mcp.Required(),
				mcp.Description("Must be true to confirm restore. With overwrite=true, all existing data on the target is destroyed. Ignored when dry_run=true.")),
			mcp.WithBoolean(paramDryRun, mcp.Description(paramDryRunDesc)),
		},
		handleInstanceBackupRestoreRequest,
	)

	return tool, profiles.CapWrite, handler
}

// validateBackupRestoreArgs validates the restore IDs, returning an error
// message or "". Shared by the real restore path and the dry-run preview.
func validateBackupRestoreArgs(linodeID, backupID, targetLinodeID int) string {
	if linodeID == 0 {
		return ErrLinodeIDRequired.Error()
	}

	if backupID == 0 {
		return ErrBackupIDRequired.Error()
	}

	if targetLinodeID == 0 {
		return "target_linode_id is required"
	}

	return ""
}

func handleInstanceBackupRestoreRequest(ctx context.Context, request *mcp.CallToolRequest, cfg *config.Config) (*mcp.CallToolResult, error) {
	overwrite := request.GetBool("overwrite", false)
	linodeID := request.GetInt("linode_id", 0)
	backupID := request.GetInt("backup_id", 0)
	targetLinodeID := request.GetInt("target_linode_id", 0)

	if IsDryRun(request) {
		if msg := validateBackupRestoreArgs(linodeID, backupID, targetLinodeID); msg != "" {
			return mcp.NewToolResultError(msg), nil
		}

		return RunDryRunPreviewDetailed(ctx, request, cfg, "linode_instance_backup_restore", httpMethodPost,
			fmt.Sprintf("/linode/instances/%d/backups/%d/restore", linodeID, backupID),
			func(ctx context.Context, c *linode.Client) (any, error) {
				return c.GetInstanceBackup(ctx, linodeID, backupID)
			},
			func(ctx context.Context, _ *linode.Client, _ any) (DryRunDetails, error) {
				return backupRestoreSideEffects(ctx, targetLinodeID, overwrite)
			})
	}

	confirmMsg := "This restores a backup to the target instance. Set confirm=true to proceed."
	if overwrite {
		confirmMsg = "This restores a backup and DESTROYS all existing disks and configs on the target instance. Set confirm=true to proceed."
	}

	if result := RequireConfirm(request, confirmMsg); result != nil {
		return result, nil
	}

	if msg := validateBackupRestoreArgs(linodeID, backupID, targetLinodeID); msg != "" {
		return mcp.NewToolResultError(msg), nil
	}

	client, err := prepareClient(request, cfg)
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	req := linode.RestoreBackupRequest{
		LinodeID:  targetLinodeID,
		Overwrite: overwrite,
	}

	if err := client.RestoreInstanceBackup(ctx, linodeID, backupID, req); err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("Failed to restore backup %d to instance %d: %v", backupID, targetLinodeID, err)), nil
	}

	response := struct {
		Message        string `json:"message"`
		BackupID       int    `json:"backup_id"`
		TargetLinodeID int    `json:"target_linode_id"`
		Overwrite      bool   `json:"overwrite"`
	}{
		Message:        fmt.Sprintf("Backup %d restore initiated to instance %d (overwrite=%t)", backupID, targetLinodeID, overwrite),
		BackupID:       backupID,
		TargetLinodeID: targetLinodeID,
		Overwrite:      overwrite,
	}

	return MarshalToolResponse(response)
}

// NewLinodeInstanceBackupsEnableTool creates a tool for enabling the backup service on a Linode instance.
func NewLinodeInstanceBackupsEnableTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newToolWithHandler(
		cfg,
		"linode_instance_backups_enable",
		"Enables the backup service for a Linode instance. "+
			"WARNING: This adds a recurring charge to your account based on the instance plan.",
		[]mcp.ToolOption{
			mcp.WithNumber("linode_id", mcp.Required(),
				mcp.Description("The ID of the Linode instance")),
			mcp.WithBoolean(paramConfirm, mcp.Required(),
				mcp.Description("Must be true to confirm enabling backups. This adds a recurring billing charge. Ignored when dry_run=true.")),
			mcp.WithBoolean(paramDryRun, mcp.Description(paramDryRunDesc)),
		},
		handleInstanceBackupsEnableRequest,
	)

	return tool, profiles.CapWrite, handler
}

func handleInstanceBackupsEnableRequest(ctx context.Context, request *mcp.CallToolRequest, cfg *config.Config) (*mcp.CallToolResult, error) {
	linodeID := request.GetInt("linode_id", 0)

	if IsDryRun(request) {
		if linodeID == 0 {
			return mcp.NewToolResultError(ErrLinodeIDRequired.Error()), nil
		}

		return RunDryRunPreview(ctx, request, cfg, "linode_instance_backups_enable", httpMethodPost,
			fmt.Sprintf("/linode/instances/%d/backups/enable", linodeID),
			func(ctx context.Context, c *linode.Client) (any, error) { return c.GetInstance(ctx, linodeID) })
	}

	if result := RequireConfirm(request, "Enabling backups adds a recurring charge to your account. Set confirm=true to proceed."); result != nil {
		return result, nil
	}

	if linodeID == 0 {
		return mcp.NewToolResultError(ErrLinodeIDRequired.Error()), nil
	}

	client, err := prepareClient(request, cfg)
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	if err := client.EnableInstanceBackups(ctx, linodeID); err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("Failed to enable backups for instance %d: %v", linodeID, err)), nil
	}

	response := struct {
		Message  string `json:"message"`
		LinodeID int    `json:"linode_id"`
	}{
		Message:  fmt.Sprintf("Backup service enabled for instance %d", linodeID),
		LinodeID: linodeID,
	}

	return MarshalToolResponse(response)
}

// NewLinodeInstanceFirewallsApplyTool creates a tool for reapplying assigned firewalls to a Linode instance.
func NewLinodeInstanceFirewallsApplyTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newToolWithHandler(
		cfg,
		"linode_instance_firewall_apply",
		"Reapplies assigned firewalls to a Linode instance. Use this if firewall assignment was not applied successfully.",
		[]mcp.ToolOption{
			mcp.WithNumber("linode_id", mcp.Required(),
				mcp.Description("The ID of the Linode instance")),
			mcp.WithBoolean(paramConfirm, mcp.Required(),
				mcp.Description("Must be true to confirm reapplying firewalls to this Linode. Ignored when dry_run=true.")),
			mcp.WithBoolean(paramDryRun, mcp.Description(paramDryRunDesc)),
		},
		handleInstanceFirewallsApplyRequest,
	)

	return tool, profiles.CapWrite, handler
}

// firewallsApplyLinodeIDFromTool validates and parses the linode_id, returning
// the ID or an error message. Shared by the real path and the dry-run preview.
func firewallsApplyLinodeIDFromTool(request *mcp.CallToolRequest) (int, string) {
	args := request.GetArguments()
	if _, exists := args["linode_id"]; !exists {
		return 0, ErrLinodeIDRequired.Error()
	}

	return optionalPaginationInt(args, "linode_id", 1, 0)
}

func handleInstanceFirewallsApplyRequest(ctx context.Context, request *mcp.CallToolRequest, cfg *config.Config) (*mcp.CallToolResult, error) {
	if IsDryRun(request) {
		linodeID, validationMessage := firewallsApplyLinodeIDFromTool(request)
		if validationMessage != "" {
			return mcp.NewToolResultError(validationMessage), nil
		}

		return RunDryRunPreview(ctx, request, cfg, "linode_instance_firewall_apply", httpMethodPost,
			fmt.Sprintf("/linode/instances/%d/firewalls/apply", linodeID),
			func(ctx context.Context, c *linode.Client) (any, error) { return c.GetInstance(ctx, linodeID) })
	}

	if result := RequireConfirm(request, "This reapplies assigned firewalls to the Linode. Set confirm=true to proceed."); result != nil {
		return result, nil
	}

	linodeID, validationMessage := firewallsApplyLinodeIDFromTool(request)
	if validationMessage != "" {
		return mcp.NewToolResultError(validationMessage), nil
	}

	client, err := prepareClient(request, cfg)
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	if err := client.ApplyInstanceFirewalls(ctx, linodeID); err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("Failed to apply firewalls for instance %d: %v", linodeID, err)), nil
	}

	response := struct {
		Message  string `json:"message"`
		LinodeID int    `json:"linode_id"`
	}{
		Message:  fmt.Sprintf("Firewall apply initiated for instance %d", linodeID),
		LinodeID: linodeID,
	}

	return MarshalToolResponse(response)
}

// NewLinodeInstanceBackupsCancelTool creates a tool for canceling the backup service on a Linode instance.
func NewLinodeInstanceBackupsCancelTool(cfg *config.Config) (mcp.Tool, profiles.Capability, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	tool, handler := newToolWithHandler(
		cfg,
		"linode_instance_backups_cancel",
		"Cancels the backup service for a Linode instance. "+
			"WARNING: This permanently deletes all existing backups for the instance."+
			" Pass dry_run=true to preview without canceling."+twoStageNote,
		[]mcp.ToolOption{
			mcp.WithNumber("linode_id", mcp.Required(),
				mcp.Description("The ID of the Linode instance")),
			mcp.WithBoolean(paramConfirm, mcp.Required(),
				mcp.Description("Must be true to confirm cancellation. All existing backups will be permanently deleted. Ignored when dry_run=true.")),
			mcp.WithBoolean(paramDryRun, mcp.Description(paramDryRunDesc)),
			mcp.WithString(paramMode, mcp.Description(paramModeDesc)),
			mcp.WithString(paramPlanID, mcp.Description(paramPlanIDDesc)),
		},
		handleInstanceBackupsCancelRequest,
	)

	return tool, profiles.CapDestroy, handler
}

func handleInstanceBackupsCancelRequest(ctx context.Context, request *mcp.CallToolRequest, cfg *config.Config) (*mcp.CallToolResult, error) {
	return RunDestructiveActionWithID(ctx, request, cfg, &DestructiveActionByID{
		ToolName:       "linode_instance_backups_cancel",
		IDParam:        paramLinodeID,
		Method:         httpMethodPost,
		PathPattern:    "/linode/instances/%d/backups/cancel",
		ConfirmMessage: "This permanently deletes all backups for the instance. Set confirm=true to proceed.",
		SuccessFormat:  "Backup service canceled for instance %d. All backups have been deleted.",
		FetchState: func(ctx context.Context, c *linode.Client, id int) (any, error) {
			return c.GetInstance(ctx, id)
		},
		Execute: func(ctx context.Context, c *linode.Client, id int) error {
			return c.CancelInstanceBackups(ctx, id)
		},
		// The plan fetches the parent instance, so reuse the Instance
		// hash-ignore list to drop its self-moving telemetry fields.
		HashIgnore: twostage.HashIgnoreFields("Instance"),
	})
}

File: python/src/linodemcp/tools/linode_instance_backups.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from mcp.types import TextContent, Tool

from linodemcp.profiles import Capability
from linodemcp.tools.helpers import (
    DRY_RUN_PROP,
    MODE_PROP,
    PARAM_DRY_RUN,
    PARAM_MODE,
    PARAM_PLAN_ID,
    PLAN_ID_PROP,
    TWO_STAGE_NOTE,
    DryRunDetails,
    execute_dry_run,
    execute_tool,
    is_dry_run,
)
from linodemcp.tools.twostage_destroy import run_two_stage_destroy
from linodemcp.twostage.hash_ignore import hash_ignore_fields

if TYPE_CHECKING:
    from linodemcp.config import Config
    from linodemcp.linode import RetryableClient


def _error_response(message: str) -> list[TextContent]:
    """Return a single-element TextContent error list."""
    return [TextContent(type="text", text=f"Error: {message}")]


_ENV_PROP: dict[str, Any] = {
    "type": "string",
    "description": "Linode environment to use (optional, defaults to 'default')",
}

_LINODE_ID_PROP: dict[str, Any] = {
    "type": "integer",
    "minimum": 1,
    "description": "The ID of the Linode instance (required)",
}

_BACKUP_ID_PROP: dict[str, Any] = {
    "type": "integer",
    "minimum": 1,
    "description": "The ID of the backup (required)",
}

_CONFIRM_PROP: dict[str, Any] = {
    "type": "boolean",
    "description": "Must be true to confirm this operation.",
}


def _parse_instance_id(
    arguments: dict[str, Any],
) -> int | list[TextContent]:
    """Parse and validate linode_id from arguments."""
    raw = arguments.get("linode_id", "")
    if not raw:
        return _error_response("linode_id is required")
    try:
        return int(raw)
    except (ValueError, TypeError):
        return _error_response("linode_id must be a valid integer")


def _parse_instance_and_backup_ids(
    arguments: dict[str, Any],
) -> tuple[int, int] | list[TextContent]:
    """Parse linode_id and backup_id from arguments."""
    iid = _parse_instance_id(arguments)
    if isinstance(iid, list):
        return iid
    raw = arguments.get("backup_id", "")
    if not raw:
        return _error_response("backup_id is required")
    try:
        backup_id = int(raw)
    except (ValueError, TypeError):
        return _error_response("backup_id must be a valid integer")
    return iid, backup_id


def create_linode_instance_backup_list_tool() -> tuple[Tool, Capability]:
    """Create the linode_instance_backup_list tool."""
    return Tool(
        name="linode_instance_backup_list",
        description=("Lists backups for a Linode instance"),
        inputSchema={
            "type": "object",
            "properties": {
                "environment": _ENV_PROP,
                "linode_id": _LINODE_ID_PROP,
            },
            "required": ["linode_id"],
        },
    ), Capability.Read


async def handle_linode_instance_backup_list(
    arguments: dict[str, Any], cfg: Config
) -> list[TextContent]:
    """Handle linode_instance_backup_list tool request."""
    iid = _parse_instance_id(arguments)
    if isinstance(iid, list):
        return iid

    async def _call(
        client: RetryableClient,
    ) -> dict[str, Any]:
        return await client.list_instance_backups(iid)

    return await execute_tool(cfg, arguments, "list instance backups", _call)


def create_linode_instance_backup_get_tool() -> tuple[Tool, Capability]:
    """Create the linode_instance_backup_get tool."""
    return Tool(
        name="linode_instance_backup_get",
        description=("Gets details of a specific backup for an instance"),
        inputSchema={
            "type": "object",
            "properties": {
                "environment": _ENV_PROP,
                "linode_id": _LINODE_ID_PROP,
                "backup_id": _BACKUP_ID_PROP,
            },
            "required": ["linode_id", "backup_id"],
        },
    ), Capability.Read


async def handle_linode_instance_backup_get(
    arguments: dict[str, Any], cfg: Config
) -> list[TextContent]:
    """Handle linode_instance_backup_get tool request."""
    ids = _parse_instance_and_backup_ids(arguments)
    if isinstance(ids, list):
        return ids
    linode_id, backup_id = ids

    async def _call(
        client: RetryableClient,
    ) -> dict[str, Any]:
        return await client.get_instance_backup(linode_id, backup_id)

    return await execute_tool(cfg, arguments, "get instance backup", _call)


def create_linode_instance_backup_create_tool() -> tuple[Tool, Capability]:
    """Create the linode_instance_backup_create tool."""
    return Tool(
        name="linode_instance_backup_create",
        description=("Creates a snapshot backup of a Linode instance"),
        inputSchema={
            "type": "object",
            "properties": {
                "environment": _ENV_PROP,
                "linode_id": _LINODE_ID_PROP,
                "label": {
                    "type": "string",
                    "description": "Label for the snapshot",
                },
                "confirm": _CONFIRM_PROP,
                PARAM_DRY_RUN: DRY_RUN_PROP,
            },
            "required": ["linode_id", "confirm"],
        },
    ), Capability.Write


async def handle_linode_instance_backup_create(
    arguments: dict[str, Any], cfg: Config
) -> list[TextContent]:
    """Handle linode_instance_backup_create tool request."""
    iid = _parse_instance_id(arguments)
    if isinstance(iid, list):
        return iid

    if is_dry_run(arguments):

        async def _fetch(client: RetryableClient) -> Any:
            return await client.get_instance(iid)

        return await execute_dry_run(
            cfg,
            arguments,
            "linode_instance_backup_create",
            "POST",
            f"/linode/instances/{iid}/backups",
            _fetch,
        )

    if not arguments.get("confirm"):
        return _error_response("Set confirm=true to proceed.")

    async def _call(
        client: RetryableClient,
    ) -> dict[str, Any]:
        return await client.create_instance_backup(iid, label=arguments.get("label"))

    return await execute_tool(cfg, arguments, "create instance backup", _call)


def create_linode_instance_backup_restore_tool() -> tuple[Tool, Capability]:
    """Create the linode_instance_backup_restore tool."""
    return Tool(
        name="linode_instance_backup_restore",
        description="Restores a backup to a Linode instance",
        inputSchema={
            "type": "object",
            "properties": {
                "environment": _ENV_PROP,
                "linode_id": _LINODE_ID_PROP,
                "backup_id": _BACKUP_ID_PROP,
                "target_linode_id": {
                    "type": "integer",
                    "description": ("Target instance ID for restore"),
                },
                "overwrite": {
                    "type": "boolean",
                    "description": ("Overwrite existing data on target"),
                },
                "confirm": _CONFIRM_PROP,
                PARAM_DRY_RUN: DRY_RUN_PROP,
            },
            "required": [
                "linode_id",
                "backup_id",
                "target_linode_id",
                "confirm",
            ],
        },
    ), Capability.Write


def _backup_restore_side_effects(target_id: int, overwrite: bool) -> DryRunDetails:
    """Phase 2 Tier A walk for backup restore. The side effect depends on the
    overwrite flag: with overwrite the target instance's existing disks and
    configs are destroyed and replaced, otherwise the backup is restored
    alongside what is already there. Args-based, no API call.
    """
    if overwrite:
        return {
            "side_effects": [
                f"All existing disks and configs on target instance {target_id} "
                "are destroyed and replaced by the backup."
            ],
            "warnings": [
                f"overwrite=true: existing data on target instance {target_id} "
                "is permanently lost."
            ],
        }
    return {
        "side_effects": [
            f"The backup is restored onto target instance {target_id}; the "
            "restore fails if its disks or configs collide."
        ]
    }


async def handle_linode_instance_backup_restore(
    arguments: dict[str, Any], cfg: Config
) -> list[TextContent]:
    """Handle linode_instance_backup_restore request."""
    ids = _parse_instance_and_backup_ids(arguments)
    if isinstance(ids, list):
        return ids
    linode_id, backup_id = ids

    raw_target = arguments.get("target_linode_id")
    if not raw_target:
        return _error_response("target_linode_id is required")
    target_linode_id = int(raw_target)

    if is_dry_run(arguments):
        overwrite = arguments.get("overwrite", False)

        async def _fetch(client: RetryableClient) -> Any:
            return await client.get_instance_backup(linode_id, backup_id)

        async def _walk(_client: RetryableClient, _state: Any) -> DryRunDetails:
            return _backup_restore_side_effects(target_linode_id, overwrite)

        return await execute_dry_run(
            cfg,
            arguments,
            "linode_instance_backup_restore",
            "POST",
            f"/linode/instances/{linode_id}/backups/{backup_id}/restore",
            _fetch,
            _walk,
        )

    if not arguments.get("confirm"):
        return _error_response("Set confirm=true to proceed.")

    overwrite = arguments.get("overwrite", False)

    async def _call(
        client: RetryableClient,
    ) -> dict[str, Any]:
        await client.restore_instance_backup(
            linode_id,
            backup_id,
            target_linode_id,
            overwrite=overwrite,
        )
        return {
            "message": (f"Backup {backup_id} restored to instance {target_linode_id}"),
            "linode_id": linode_id,
            "backup_id": backup_id,
        }

    return await execute_tool(cfg, arguments, "restore instance backup", _call)


def create_linode_instance_backups_enable_tool() -> tuple[Tool, Capability]:
    """Create the linode_instance_backups_enable tool."""
    return Tool(
        name="linode_instance_backups_enable",
        description=("Enables backups for a Linode instance (billing charges apply)"),
        inputSchema={
            "type": "object",
            "properties": {
                "environment": _ENV_PROP,
                "linode_id": _LINODE_ID_PROP,
                "confirm": _CONFIRM_PROP,
                PARAM_DRY_RUN: DRY_RUN_PROP,
            },
            "required": ["linode_id", "confirm"],
        },
    ), Capability.Write


async def handle_linode_instance_backups_enable(
    arguments: dict[str, Any], cfg: Config
) -> list[TextContent]:
    """Handle linode_instance_backups_enable request."""
    iid = _parse_instance_id(arguments)
    if isinstance(iid, list):
        return iid

    if is_dry_run(arguments):

        async def _fetch(client: RetryableClient) -> Any:
            return await client.get_instance(iid)

        return await execute_dry_run(
            cfg,
            arguments,
            "linode_instance_backups_enable",
            "POST",
            f"/linode/instances/{iid}/backups/enable",
            _fetch,
        )

    if not arguments.get("confirm"):
        return _error_response("Set confirm=true to proceed.")

    async def _call(
        client: RetryableClient,
    ) -> dict[str, Any]:
        await client.enable_instance_backups(iid)
        return {
            "message": (f"Backups enabled for instance {iid}"),
            "linode_id": iid,
        }

    return await execute_tool(cfg, arguments, "enable instance backups", _call)


def create_linode_instance_backups_cancel_tool() -> tuple[Tool, Capability]:
    """Create the linode_instance_backups_cancel tool."""
    return Tool(
        name="linode_instance_backups_cancel",
        description=(
            "Cancels backups for a Linode instance."
            " All existing backups will be deleted."
            " Pass dry_run=true to preview without canceling."
        )
        + TWO_STAGE_NOTE,
        inputSchema={
            "type": "object",
            "properties": {
                "environment": _ENV_PROP,
                "linode_id": _LINODE_ID_PROP,
                "confirm": {
                    "type": "boolean",
                    "description": (
                        "Must be true to confirm. Existing backups will be deleted."
                        " Ignored when dry_run=true."
                    ),
                },
                PARAM_DRY_RUN: DRY_RUN_PROP,
                PARAM_MODE: MODE_PROP,
                PARAM_PLAN_ID: PLAN_ID_PROP,
            },
            "required": ["linode_id", "confirm"],
        },
    ), Capability.Destroy


async def _instance_backups_cancel_two_stage(
    arguments: dict[str, Any], cfg: Config, iid: int
) -> list[TextContent] | None:
    """Run the plan/apply flow when mode is plan/apply, else None to fall through."""
    if arguments.get("mode") not in ("plan", "apply"):
        return None

    async def _ts_fetch(client: RetryableClient) -> Any:
        return await client.get_instance(iid)

    async def _ts_call(client: RetryableClient) -> dict[str, Any]:
        await client.cancel_instance_backups(iid)
        return {
            "message": f"Backups cancelled for instance {iid}",
            "linode_id": iid,
        }

    return await run_two_stage_destroy(
        cfg,
        arguments,
        tool_name="linode_instance_backups_cancel",
        method="POST",
        path=f"/linode/instances/{iid}/backups/cancel",
        fetch_state=_ts_fetch,
        execute=_ts_call,
        hash_ignore=hash_ignore_fields("Instance"),
    )


async def handle_linode_instance_backups_cancel(
    arguments: dict[str, Any], cfg: Config
) -> list[TextContent]:
    """Handle linode_instance_backups_cancel request."""
    iid = _parse_instance_id(arguments)
    if isinstance(iid, list):
        return iid

    two_stage = await _instance_backups_cancel_two_stage(arguments, cfg, iid)
    if two_stage is not None:
        return two_stage

    if is_dry_run(arguments):

        async def _fetch(client: RetryableClient) -> Any:
            return await client.get_instance(iid)

        return await execute_dry_run(
            cfg,
            arguments,
            "linode_instance_backups_cancel",
            "POST",
            f"/linode/instances/{iid}/backups/cancel",
            _fetch,
        )

    confirm = arguments.get("confirm", False)
    if not confirm:
        return _error_response("This is destructive. Set confirm=true to proceed.")

    async def _call(
        client: RetryableClient,
    ) -> dict[str, Any]:
        await client.cancel_instance_backups(iid)
        return {
            "message": (f"Backups cancelled for instance {iid}"),
            "linode_id": iid,
        }

    return await execute_tool(cfg, arguments, "cancel instance backups", _call)
